src/puppetlabs/trapperkeeper/internal.clj

```diff
@@ -193,15 +201,37 @@
       (log/errorf t "Error during service %s!!!" lifecycle-fn-name)
       (throw t))))
 
-(def ^:private sighup-lock (Object.))
+(schema/defn ^:always-validate initialize-lifecycle-worker :- (schema/protocol async-prot/Channel)
+  "Initializes a 'worker' which will listen for lifecycle-related tasks and perform
+  them on a background thread, to ensure that we aren't executing multiple lifecycle
+  tasks simultaneously."
+  [lifecycle-chan :- (schema/protocol async-prot/Channel)
+   shutdown-reason-promise :- IDeref]
+  (async/go (try
+              (loop []
+                (let [task (async/<! lifecycle-chan)]
+                  (if (= task :shutdown)
+                    (do
+                      (log/debug "Received shutdown command, lifecycle worker exiting.")
+                      :done)
+                    (do
+                      (log/debug "Lifecycle worker executing lifecycle task.")
+                      (task)
+                      (log/debug "Lifecycle worker completed lifecycle task; awaiting next task.")
+                      (recur)))))
+              (catch Exception e
+                (deliver shutdown-reason-promise
+                         {:cause :service-error
+                          :error e})))))
 
 (defn restart-tk-apps
   "Call restart on all tk apps."
   [apps]
   (log/debug "SIGHUP handler restarting TK apps.")
-  (locking sighup-lock
-    (doseq [app apps]
-      (a/restart app))))
+  (doseq [app apps]
+    (let [{:keys [lifecycle-channel]} @(a/app-context app)
+          restart-fn #(a/restart app)]
+      (async/>!! lifecycle-channel restart-fn))))
```

Comment:
In this implementation each app has its own lifecycle channel, and they could actually be restarted in parallel.  I don't feel too strongly about it but I think I'd rather keep it that way until we have a reason not to; it's not really visible in the API so I think we could change it later if we needed.